### DISCUSSION TITLE
Multi tenant filament app and image uploads

### DISCUSSION THREAD

**Discussion by maxroby** [filamentphp/filament#5635] [Help]
Hi all, I'm starting a demo with stancl/tenancy and filament for tenant apps, but I have a problem with the image file upload field included.

I'm using
```
FileUpload::make('image')
                    ->image()->enableOpen()
                    ->directory('assets'),
```
in my admin panel, having the image correctly uploaded to storage/tenantacme/app/public/assets directory and shown in the preview.

But when I return to edit the item, the image is not displayed:

![immagine](https://user-images.githubusercontent.com/2692822/217767963-2582f9a8-4f90-44b9-9e27-12ae0c3e9028.png)

This happens because it's trying to load the image from _localhost/storage/assets/_, ignoring the tenant domain and folder.

Thanks for all your hints, I'll be glad to give more information or try to fix.

Regards,
Roberto

**Comment by danharrin** [comment]
This is a disk misconfiguration, it's not using the correct domain. Not something we control using Filament, it's a Laravel filesystems thing. Try setting the `config('app.url')` dynamically

**Comment by maxroby** [comment]
Hi and thanks for the response!
I tried changing the _app.url_ config, but it doesn't work:
URL is still referring to _/storage/assets/_ while it has to be _/tenancy/assets/_ like returned by the asset() Laravel method.
Is there a way to force Filament to use this function?

I've tried to show the field in a table using a custom view with _asset($getState())_ and it works.

**Comment by maxroby** [comment]
Found that changing line 128 of BaseFileUpload.php with:
`return asset($file);`
makes it work for edit page and line 150 of ImageColumn.php to:
`return asset($state);`
for the table.
Is it a correct solution?

**Reply by danharrin** [reply]
If the filesystem disk is correctly configured to generate the URL, then you won't need to hack it like this. It's not a solution to change vendor files as they won't survive Composer updates.

I don't know how to configure the disk with a dynamic URL, that's something you will need to research in the wider Laravel community.

**Comment by stefanolissa** [comment]
Uhm, I'm following this discussion since I have a similar problem. `FileSystem::url()` returns the URL of a resource, not always an HTTP URL but even an FTP URL and like. I can understand that from the Laravel implementation of those classes. 

Using `asset()` or `url()` helpers, means using an UrlGenerator which can be "replaced" by other components to create a different URL that can be a proxy to the storage, introduce validation, protection, and so on.

Just my two cents.

**Reply by danharrin** [reply]
Exactly, hence why Filament strictly reads from the filesystem adapters and never tries to generate its own URLs

**Comment by Javdu10** [comment]
@maxroby this might help you : https://github.com/archtechx/tenancy/pull/909

You have two solution, use an S3 compatible disk, or implement a piece of this PR which is for tenancy V4

**Comment by danihidayatx** [comment]
After a full day of headaches, here's the solution I used to make FileUpload work correctly with tenant-aware storage and show the correct image preview in Filament v3 using Stancl Tenancy.

---

1. In `config/filesystems.php` add a `tenant` disk:

'tenant' => [
    'driver' => 'local',
    'root' => storage_path('app/public'),
    'url' => '/tenancy/assets',
    'visibility' => 'public',
    'throw' => false,
],

---

2. In `config/tenancy.php`, update the filesystem config:

'filesystem' => [
    'disks' => ['tenant'],
    'root_override' => [
        'tenant' => '%storage_path%/app/public',
    ],
    'suffix_storage_path' => true,
],

---

3. Use it in Filament form:

FileUpload::make('logo')
    ->disk('tenant')
    ->directory('studios')
    ->image()
    ->visibility('public')

---

4. For the table preview, use:

Tables\Columns\ImageColumn::make('logo')
                    ->getStateUsing(fn ($record) => $record->logo ? tenant_asset($record->logo) : null)
                    ->circular(),

---

✅ Result: Images are stored in tenant-specific storage and correctly displayed via `/tenancy/assets/...`

⚠️ Limitation: Form preview still uses Storage::url() and can’t be overridden in Filament v3 (no more getUploadedFileUrlUsing). So preview during editing may break, but upload and listing work fine.

Hope this helps someone and saves hours of trial and error.

**Reply by xarmengol** [reply]
Thanks a lot for your reply,  I finally got it working!

**Reply by marciomdrs2002** [reply]
thx man!!! 1 night lost on it
